containers/src/Data/IntMap/Internal.hs

```diff
@@ -2301,37 +2301,37 @@ deleteFindMax = fromMaybe (error "deleteFindMax: empty map has no maximal elemen
 deleteFindMin :: IntMap a -> ((Key, a), IntMap a)
 deleteFindMin = fromMaybe (error "deleteFindMin: empty map has no minimal element") . minViewWithKey
 
+lookupMinSure :: IntMap a -> (Key, a)
+lookupMinSure (Tip k v)   = (k,v)
+lookupMinSure (Bin _ l _) = lookupMinSure l
+lookupMinSure Nil         = error "lookupMinSure Nil"
+
 -- | \(O(\min(n,W))\). The minimal key of the map. Returns 'Nothing' if the map is empty.
 lookupMin :: IntMap a -> Maybe (Key, a)
-lookupMin Nil = Nothing
-lookupMin (Tip k v) = Just (k,v)
-lookupMin (Bin p l r)
-  | signBranch p = go r
-  | otherwise = go l
-    where go (Tip k v)    = Just (k,v)
-          go (Bin _ l' _) = go l'
-          go Nil          = Nothing
+lookupMin Nil         = Nothing
+lookupMin (Tip k v)   = Just (k,v)
+lookupMin (Bin p l r) = Just $! lookupMinSure (if signBranch p then r else l)
+{-# INLINE lookupMin #-}
```

Comment:
This all deserves a brief note explaining design choices.

Reply:
Done. There was already a related note in `Data.Set.Internal`, updated that.